src/client/ua_client_subscriptions.c

```diff
@@ -947,39 +947,64 @@ UA_MonitoredItem_change_clientHandle(void *data, UA_Client_MonitoredItem *mon) {
     return NULL;
 }
 
+static void
+UA_MonitoredItem_change_clientHandle(UA_Client_Subscription *sub,
+                                     UA_ModifyMonitoredItemsRequest *request) {
+    for(size_t i = 0; i < request->itemsToModifySize; ++i) {
+        ZIP_ITER(MonitorItemsTree, &sub->monitoredItems,
+                 UA_MonitoredItem_change_clientHandle_wrapper,
+                 &request->itemsToModify[i]);
+    }
+}
+
 UA_ModifyMonitoredItemsResponse
 UA_Client_MonitoredItems_modify(UA_Client *client,
                                 const UA_ModifyMonitoredItemsRequest request) {
     UA_ModifyMonitoredItemsResponse response;
+    UA_ModifyMonitoredItemsResponse_init(&response);
 
     UA_LOCK(&client->clientMutex);
-    UA_Client_Subscription *sub;
-    LIST_FOREACH(sub, &client->subscriptions, listEntry) {
-        if (sub->subscriptionId == request.subscriptionId)
-            break;
-    }
-
+    UA_Client_Subscription *sub = findSubscription(client, request.subscriptionId);
+    UA_UNLOCK(&client->clientMutex);
     if(!sub) {
-        UA_ModifyMonitoredItemsResponse_init(&response);
         response.responseHeader.serviceResult = UA_STATUSCODE_BADSUBSCRIPTIONIDINVALID;
-        UA_UNLOCK(&client->clientMutex);
         return response;
     }
 
     UA_ModifyMonitoredItemsRequest modifiedRequest;
     UA_ModifyMonitoredItemsRequest_copy(&request, &modifiedRequest);
+    UA_MonitoredItem_change_clientHandle(sub, &modifiedRequest);
+
+    __UA_Client_Service(client, &modifiedRequest,
+                     &UA_TYPES[UA_TYPES_MODIFYMONITOREDITEMSREQUEST], &response,
+                     &UA_TYPES[UA_TYPES_MODIFYMONITOREDITEMSRESPONSE]);
 
-    for(size_t i = 0; i < modifiedRequest.itemsToModifySize; ++i) {
-        ZIP_ITER(MonitorItemsTree, &sub->monitoredItems,
-                 UA_MonitoredItem_change_clientHandle, &modifiedRequest.itemsToModify[i]);
-        __Client_Service(client, &modifiedRequest,
-                         &UA_TYPES[UA_TYPES_MODIFYMONITOREDITEMSREQUEST], &response,
-                         &UA_TYPES[UA_TYPES_MODIFYMONITOREDITEMSRESPONSE]);
-    }
     UA_ModifyMonitoredItemsRequest_clear(&modifiedRequest);
+    return response;
+}
 
+UA_StatusCode
+UA_Client_MonitoredItems_modify_async(UA_Client *client,
+                                      const UA_ModifyMonitoredItemsRequest request,
+                                      UA_ClientAsyncServiceCallback callback,
+                                      void *userdata, UA_UInt32 *requestId) {
+    UA_LOCK(&client->clientMutex);
+    UA_Client_Subscription *sub = findSubscription(client, request.subscriptionId);
```

Comment:
Also don’t unlock here in the good case.

Reply:
Fixed in https://github.com/open62541/open62541/pull/6297/commits/0552466662d387814e912e3ce22114befea5b305